apps/ejabberd/src/mod_mam.erl

```diff
@@ -281,23 +285,28 @@ filter_packet(drop) ->
     drop;
 filter_packet({From, To=#jid{luser=LUser, lserver=LServer}, Packet}) ->
     ?DEBUG("Receive packet~n    from ~p ~n    to ~p~n    packet ~p.",
-              [From, To, Packet]),
-    Packet2 =
-    case ejabberd_users:is_user_exists(LUser, LServer) of
-    false -> Packet;
-    true ->
-        case {handle_package(incoming, true, To, From, From, Packet),
-              add_archived_element()} of
-            {undefined, _} -> Packet;
-            {_, false} -> Packet;
-            {MessID, true} ->
-                ?DEBUG("Archived incoming ~p", [MessID]),
-                BareTo = jid:to_binary(jid:to_bare(To)),
-                replace_archived_elem(BareTo, MessID, Packet)
-        end
-    end,
-    {From, To, Packet2}.
+           [From, To, Packet]),
+    {AmpEvent, PacketAfterArchive} =
+        case ejabberd_users:is_user_exists(LUser, LServer) of
+            false ->
+                {failed, Packet};
+            true ->
+                PacketWithoutAmp = mod_amp:strip_amp_el_from_request(Packet),
+                case {process_incoming_packet(From, To, PacketWithoutAmp),
```

Comment:
How about moving this case to a separate function? I think this would improve readability.